File: Frontend/src/services/gameService.js
import { sendPacket } from "../network/dispatcher";
import { OPCODE } from "../network/opcode";

/**
 * Start a match in the specified room
 * @param {number} roomId - The room ID
 */
export function startGame(roomId) {
    if (!roomId || roomId <= 0) {
        console.error("[gameService] Invalid room ID:", roomId);
        return;
    }

    console.log(`[gameService] Starting game in room ${roomId}...`);

    // Build binary payload: uint32_t room_id (Big Endian)
    const payload = new ArrayBuffer(4);
    const view = new DataView(payload);
    view.setUint32(0, roomId, false); // Big Endian (network byte order)

    // Send packet (fire and forget)
    // Response (RES_GAME_STARTED) or Notification (NTF_GAME_START) 
    // should be handled by listeners in the UI or global handlers.
    sendPacket(OPCODE.CMD_START_GAME || 0x0300, payload);
}

/**
 * Leave/end the current match
 * @param {number} matchId - The match ID
 * @returns {Promise<{success: boolean}>}
 */
export async function endMatch(matchId) {
    console.log(`[gameService] Ending match ${matchId}...`);

    // TODO: Implement end match logic
    // For now, return placeholder
    return { success: false, error: "Not implemented" };
}

File: app/client/flutter_app/lib/services/game_service.dart
import 'dart:async';
import 'dart:typed_data';
import '../network/tcp_client.dart';
import '../network/dispatcher.dart';
import '../core/command.dart';
import '../core/protocol.dart';

class GameService {
  final TcpClient client;
  final Dispatcher dispatcher;

  GameService(this.client, this.dispatcher);

  /// Start a match in the specified room
  /// Backend will broadcast NTF_GAME_START to all players (no direct response)
  Future<Map<String, dynamic>> startGame(int roomId) async {
    final buffer = ByteData(4);
    buffer.setUint32(0, roomId, Endian.big);

    try {
      // Send command without waiting for response
      // Backend will broadcast NTF_GAME_START to all room members
      final packet = Protocol.buildPacket(
        command: Command.startGame,
        seqNum: 0,
        payload: buffer.buffer.asUint8List(),
      );
      
      client.send(packet);
      
      print("[GameService] Start game request sent for room $roomId");
      
      // Success means command was sent, actual game start is confirmed via NTF_GAME_START
      return {"success": true};
    } catch (e) {
      return {"success": false, "error": e.toString()};
    }
  }

  Future<Map<String, dynamic>> forfeit(int matchId) async {
    final buffer = ByteData(4);
    buffer.setUint32(0, matchId);

    try {
      final response = await client.request(Command.forfeit, payload: buffer.buffer.asUint8List());
      if (response.command == Command.resSuccess) {
        return {"success": true};
      } else {
        return {"success": false, "error": "Unexpected response: ${response.command}"};
      }
    } catch (e) {
      return {"success": false, "error": e.toString()};
    }
  }

  Future<void> endMatch(int matchId) async {
      // TODO: Implement end match
      print("[GameService] End match $matchId not implemented");
  }
}
